src/coreclr/src/gc/gc.cpp

```diff
@@ -5569,41 +5566,58 @@ bool gc_heap::virtual_commit (void* address, size_t size, int h_number, bool* ha
                               virtual_alloc_commit_for_heap (address, size, h_number)) :
                               GCToOSInterface::VirtualCommit(address, size));
 
-    if (!commit_succeeded_p && heap_hard_limit)
+    if (!commit_succeeded_p)
     {
-        check_commit_cs.Enter();
-        dprintf (1, ("commit failed, updating %Id to %Id",
-                current_total_committed, (current_total_committed - size)));
-        current_total_committed -= size;
-        if (h_number < 0)
-            current_total_committed_bookkeeping -= size;
+        if (heap_hard_limit)
+        {
+            check_commit_cs.Enter();
+            committed_by_oh[oh] -= size;
+            
+            dprintf (1, ("commit failed, updating %Id to %Id",
+                    current_total_committed, (current_total_committed - size)));
+            current_total_committed -= size;
+            if (h_number < 0)
+                current_total_committed_bookkeeping -= size;
 
-        check_commit_cs.Leave();
+            check_commit_cs.Leave();
+        }
     }
-
     return commit_succeeded_p;
 }
 
-bool gc_heap::virtual_decommit (void* address, size_t size, int h_number)
+bool gc_heap::virtual_decommit (void* address, size_t size, gc_oh_num oh, int h_number)
 {
 #ifndef HOST_64BIT
     assert (heap_hard_limit == 0);
 #endif //!HOST_64BIT
 
     bool decommit_succeeded_p = GCToOSInterface::VirtualDecommit (address, size);
 
-    if (decommit_succeeded_p && heap_hard_limit)
+    if (decommit_succeeded_p)
     {
-        check_commit_cs.Enter();
-        current_total_committed -= size;
-        if (h_number < 0)
-            current_total_committed_bookkeeping -= size;
-        check_commit_cs.Leave();
+        if (heap_hard_limit)
+        {
+            check_commit_cs.Enter();
+            committed_by_oh[oh] -= size;
+            current_total_committed -= size;
+            if (h_number < 0)
+                current_total_committed_bookkeeping -= size;
+            check_commit_cs.Leave();
+        }
     }
 
     return decommit_succeeded_p;
 }
 
+void gc_heap::virtual_free (void* add, size_t allocated_size, heap_segment* sg)
+{
+    assert(!heap_hard_limit);
+    GCToOSInterface::VirtualRelease (add, allocated_size);
```

Comment:
>GCToOSInterface::VirtualRelease (add, allocated_size); [](start = 4, length = 54)

this could actually fail, could take this opportunity to fix it especially now you want to track things closely